tests/test_convert.py

```diff
@@ -840,3 +973,154 @@ def test_alias_rejects_bad_types():
+    with pytest.raises(ConversionError, match="Virtual qubit declared in physical addressing mode"):
+        parse(source)
+
+
+def test_reject_mixed_addressing_mode_virtual_register():
+    source = """
+    reset $0;
+    qubit[3] q;
+    """
+    with pytest.raises(ConversionError, match="Virtual qubit declared in physical addressing mode"):
+        parse(source)
+
+
+def test_reject_hardware_qubit_in_gate_body_1():
+    source = """
+        include 'stdgates.inc';
+
+        gate my_gate q {
+           h $0;
+        }
+    """
+    with pytest.raises(ConversionError, match="hardware qubits not allowed in gate definitions"):
+        parse(source)
+
+
+def test_reject_hardware_qubit_in_gate_body_2():
+    source = """
+        include 'stdgates.inc';
+
+        h $0;
+        gate my_gate q {
+           h $0;
+        }
+    """
+    with pytest.raises(ConversionError, match="hardware qubits not allowed in gate definitions"):
+        parse(source)
+
+
+def _make_layout(hw_qubit_numbers):
+    qr = QuantumRegister(len(hw_qubit_numbers), "qr")
+    return TranspileLayout(
+        Layout.from_intlist(hw_qubit_numbers, qr),
+        dict(zip(qr, hw_qubit_numbers)),
+    )
+
+
+# pylint: disable=protected-access
+def _add_layout(circuit, hw_qubit_numbers):
+    circuit._layout = _make_layout(hw_qubit_numbers)
+
+
+# pylint: disable=protected-access
+def test_layout_for_hardware_qubits():
+    source = """
+        reset $99;
+    """
+    qc = parse(source)
+    expected = QuantumCircuit([Qubit()])
+    expected.reset(0)
+    _add_layout(expected, [99])
+
+    assert qc == expected
+    assert qc._layout == expected._layout
+
+    expected2 = QuantumCircuit([Qubit()])
+    expected2.reset(0)
+    _add_layout(expected2, [0])
+
+    assert qc._layout != expected2._layout
+
+
+def test_hardware_qubit_local_scope():
+    source = """
+    include "stdgates.inc";
+
+    bit[2] mid;
+
+    reset $100;
+    reset $101;
+    while (mid == "00") {
+        h $100;
+        h $101;
+        mid[0] = measure $100;
+        mid[1] = measure $101;
+     }
+    """
+
+    qc = parse(source)
+    layout = _make_layout([100, 101])
+    assert qc._layout == layout
+
+
+def test_hardware_qubit_nested_scope():
+    source = """
+    include "stdgates.inc";
+
+    bit[2] mid;
+
+    reset $100;
+    reset $101;
+
+    if (mid[0]) {
+        while (mid == "00") {
+            h $100;
+            h $101;
+            mid[0] = measure $100;
+            mid[1] = measure $101;
+         }
+    }
+    """
+
+    qc = parse(source)
+    layout = _make_layout([100, 101])
+    assert qc._layout == layout
+
+
+def test_hardware_qubit_first_seen_in_local_scope():
+    source = """
+    include "stdgates.inc";
+
+    bit[2] mid;
+
+    while (mid == "00") {
+        h $100;
+        h $101;
+        mid[0] = measure $100;
+        mid[1] = measure $101;
+     }
+    """
+
+    qc = parse(source)
+    layout = _make_layout([100, 101])
+    assert qc._layout == layout
```

Comment:
I think we also need to test that these hardware qubits can then be used outside the local scope they're first seen in, and that that produces expected outputs (i.e. still refers to the same qubit) and ideally the lookup doesn't involve any copies.

As a corollary of that: can we just that the addressing mode getting fixed to "hardware" within a nested scope causes the expected failures if a virtual qubit is first declared _after_ (e.g. the program `bit[2] mid; while (mid == "00") { h $0; } qubit q;` should raise an error).

Reply:
I agree these are good tests to have. I'm not sure what you mean by "lookup doesn't involve any copies", and furthermore what you'd like to test about it. I'll add this test first. Then implement your alternative design above for `SymbolTable`.

Reply:
I meant something along the lines that "seeing `$0` in a new scope doesn't cause a new entry to be added to the symbol table; it should already be there", but I didn't explain it very well.  It's not actually super important to try and test that separately, I think - it's kind of an implementation detail.

Reply:
Test exception raised if: Fix HW addressing mode in local scope. Then declare virtual qubit.
Done in efa1e031ab

Reply:
Test hardware qubit use after introduction in a previous local scope
Done in 855dc138303